Tasks/Gradle/gradletask.ts

```diff
@@ -131,36 +131,47 @@ if (specifiedJavaHome) {
     tl.debug('Set JAVA_HOME to ' + specifiedJavaHome);
     process.env['JAVA_HOME'] = specifiedJavaHome;
 }
- 
+
 /* Actual execution of Build and further flows*/
 async function execBuild() {
     await execEnableCodeCoverage();
     if (reportingTaskName && reportingTaskName != "") {
         gb.arg(reportingTaskName);
     }
-    
+
     enableSonarQubeAnalysis();
     var gradleResult;
+    var statusFailed = false;
     gb.exec()
         .then(function (code) {
             gradleResult = code;
             tl.debug(`exit code: ${code}`);
-            publishTestResults(publishJUnitResults, testResultsFiles);
-            publishCodeCoverage(isCodeCoverageOpted);
             return processCodeAnalysisResults();
         })
         .then(() => {
             tl.debug(`Gradle result: ${gradleResult}`);
-            if (gradleResult === 0) {
-                tl.setResult(tl.TaskResult.Succeeded, "Build succeeded.");
-            } else {
-                tl.setResult(tl.TaskResult.Failed, "Build failed.");
-            }
+            return Q.resolve("Success");
         })
         .fail(function (err) {
             console.error(err);
             tl.debug('taskRunner fail');
-            tl.setResult(tl.TaskResult.Failed, err);
+            gradleResult = -1;
+            statusFailed = true;
+            return Q.resolve(err);
+        })
+        .then(function (resp) {
+            // We should always publish test results and code coverage
+            publishTestResults(publishJUnitResults, testResultsFiles);
+            publishCodeCoverage(isCodeCoverageOpted);
+
+            if (gradleResult === 0) {
+                tl.setResult(tl.TaskResult.Succeeded, "Build succeeded.");
+            } else if (gradleResult === -1 && statusFailed === true) {
```

Comment:
Shouldn't this `if` statement use `||` instead of `&&`?

`} else if (gradleResult === -1 || statusFailed === true) {`